zephyr/test/vmh.c

```diff
@@ -17,8 +18,258 @@
+{
+	struct vmh_heap *heap =
+	vmh_init_heap(NULL, MEM_REG_ATTR_CORE_HEAP, 0, allocating_continuously);
+
+	zassert_not_null(heap, "Heap initialization failed");
+
+	test_vmh_alloc_free_no_check(heap, 512, true);
+	test_vmh_alloc_free_no_check(heap, 2048, true);
+	test_vmh_alloc_free_no_check(heap, sizeof(int), false);
+	test_vmh_alloc_free_no_check(heap, 0, false);
+
+	test_vmh_alloc_free_check(heap, 512, true);
+	test_vmh_alloc_free_check(heap, 2048, true);
+	test_vmh_alloc_free_check(heap, sizeof(int), false);
+	test_vmh_alloc_free_check(heap, 0, false);
+
+	int ret = vmh_free_heap(heap);
+
+	zassert_equal(ret, 0, "Failed to free heap");
+
+	/* Could add tests with configs for heaps*/
+}
+
+/* Test case for vmh_alloc and vmh_free with and without config */
+static void test_heap_creation(void)
+{
+	test_vmh_init_and_free_heap(MEM_REG_ATTR_CORE_HEAP, NULL, 0, false, true);
+
+	/* Try to setup with pre defined heap config */
+	struct vmh_heap_config config;
+
+	config.block_bundles_table[0].block_size = 4;
+
+	config.block_bundles_table[0].number_of_blocks = 80;
+
+	config.block_bundles_table[1].block_size = 4;
+
+	config.block_bundles_table[1].number_of_blocks = 80;
+
+	test_vmh_init_and_free_heap(MEM_REG_ATTR_CORE_HEAP, &config, 0, false, true);
+}
+
+/* Test case for alloc/free on configured heap */
+static void test_alloc_on_configured_heap(bool allocating_continuously)
+{
+
+	/* Try to setup with pre defined heap config */
+	struct vmh_heap_config config;
+
+	config.block_bundles_table[0].block_size = 20;
+
+	config.block_bundles_table[0].number_of_blocks = 80;
+
+	/* Create continuous allocation heap for success test */
+	struct vmh_heap *heap =
+	vmh_init_heap(NULL, MEM_REG_ATTR_CORE_HEAP, 0, allocating_continuously);
+
+	/* Will succeed on continuous and fail with single block alloc */
+	test_vmh_alloc_free_check(heap, 512, allocating_continuously);
+
+	int ret = vmh_free_heap(heap);
+
+	zassert_equal(ret, 0, "Failed to free heap");
+}
+
+/* Test cases for initializing heaps on all available regions */
+static void test_vmh_init_all_heaps(void)
+{
+	int num_regions = CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT;
+	int i;
+	const struct sys_mm_drv_region *virtual_memory_region =
+	sys_mm_drv_query_memory_regions();
+
+	/* Test initializing all types of heaps */
+	for (i = 0; i < num_regions; i++) {
+
+		/* Test without continuous allocation */
+		test_vmh_init_and_free_heap(virtual_memory_region[i].attr, NULL, 0, false,
+		true);
+
+		/* Test if it fails when heap already exists */
+		test_vmh_init_and_free_heap(virtual_memory_region[i].attr, NULL, 0, true,
+		false);
+	}
+}
 
 ZTEST(sof_boot, virtual_memory_heap)
 {
+	test_heap_creation();
+	test_vmh_init_all_heaps();
+	test_alloc_on_configured_heap(true);
+	test_alloc_on_configured_heap(false);
+	test_vmh_alloc_free(true);
+	test_vmh_alloc_free(false);
+	test_vmh_alloc_multiple_times(true);
+	test_vmh_alloc_multiple_times(false);
```

Comment:
This is quite a bit of testing, taking into account that many of these functions test multiple allocations too. Do we know how long this takes? I'm not sure we need that much testing - I'd basically test each (or most) configuration once - once in contiguous mode, once without, once with success, once with failure. Something rather lightweight for regular debug-overlay testing, possibly add a new Kconfig for extensive VMH testing. So do all these tests now pass? Can we re-enable boot-time testing by default?

Reply:
Overall I assumed that this boot test feature is not intended to be enabled in the productized version. So we will only use the CONFIG_SOF_BOOT_TEST flag for testing only and are not that much impacted by test time.

However, for allocator testing, this is quite concise. Should not take that much time - will have better data on it as soon as I clean up all the issues those tests found.

Amount of tests: This test approach allows for clear identification of the issues. I tried to add tests for each case I encountered. For example, multiple allocations failed in my internal testing during development so now I want it tested extensively. Writing data to the allocated space also checks for correctly assigned physical pages.

Can we enable boot testing? No - I am still working out the issues in vmh API #8772 Once this is done It can be pushed.

Reply:
Good, that this helps you find issues with the VMH! Maybe we should first have all the issues fixed and then merge and enable this?

Reply:
I am working on it so I see no problem In pushing both fixes and this at the same time.
Linked pr is for fixes I will do.

Reply:
It does not matter how long the tests take to run as they should not be used in a production mode. i.e. we are not going to add latency to audio use case.
CI should be able to build this Kconfig and all tests should run for all subsystems.

Reply:
I'd prefer to enable this in debug overlay, so this would get tested automatically with the current infra we have. That would require the tests to pass. Otherwise the code will very easily get broken (like it has when we temporarily disabled it in the debug overlay).